### ISSUE TITLE
Future compatibility

### ISSUE THREAD

**Issue by theiding** [sbozarth/homebridge-lc7001#1]
@sbozarth, thank you so much for sharing your work - I really appreciate it. Could you document (very briefly) the history of your integration with the LC7001, i.e. how did you create lc7001.js? Is it based on reverse engineering what the Legrand Lighting Control app sends over the network? Or is any of the integration officially published/supported by Legrand? Could Legrand break any of it in the future without notice?
Thank you again for your great work!
Mark

**Comment by sbozarth** [comment]
I reverse engineered the protocol by packet sniffing the exchange between the iOS app and the LC7001. Made much, much easier by the fact that it is not encrypted. Turns out that it’s just a straight TCP socket exchanging JSON in plain text.

So, yes, it could break at any time. The good news is that the LC7001 does not auto-update as best as I can tell. So if you never apply an update, you’re good. I applied one while holding my breath. My LC7001 is working on version 3.1.1-8-g0425.

Sean

**Comment by theiding** [comment]
Awesome - thanks for clarifying. I am looking into integrating the lc7001 with Openhab (i.e. implementing an Openhab binding just like you implemented a homebridge plugin). 
Looks like Artik is now a feasible option for a stable API into the lc7001. If that doesn't work I will come back and leverage your lc7001.js.

**Comment by theiding** [comment]
I almost completed the Samsung ARTIK Cloud binding for Openhab. And then I saw Samsung's announcement that they are shutting down ARTIK Cloud. So much for a stable API - arrrrrghhhh! lc7001.js is the most stable API to date after all!

**Comment by mkrapivner** [comment]
Actually, it looks like there’s a very low chance of it breaking, because some of Legrand’s integrations with other home automation controllers available on their website (Control4, RTI, etc. ) date back to 2015 and haven’t been updated since. Thanks Sean for researching this, will need to add it to SmartThings hub somehow, now that ARTIK cloud is shut down.

**Comment by sbozarth** [comment]
There is a new iOS app, version 4.0. There is also a new firmware, 4:0:0-269. I have not applied the firmware since my system is working fine and I didn’t use ARTIK. Not sure I want to be the guinea pig. Definitely too busy at the moment to start a side project.

The iOS notes mention a “transition to Legrand cloud.” Say it will still work with Alexa and Google assistant via skill.

Sean

**Comment by ghost** [comment]
Dang, I wish I would have read this before upgrading my firmware...I guess that makes me the guinea pig. 

As of today, I am getting 2 new errors in my homekit console, although it does in face still seem to be actually working. (Odd)

Added the 2 errors as open issues

- Unknown LC7001 initiated object! - As of today (when I upgraded to firmware 4.0.1-28
- Unhandled service: Eliot Errors - The other day when I upgraded to firmware 4.0.0-269

**Comment by sbozarth** [comment]
The good news is that you've cleared the first major hurdle. If your LC7001 has the new firmware and the plugin can still contact it and throw the first "error," that is fantastic! That means none of the connection logic has changed.

Unknown LC7001 initiated object! -- This is not really an error. If I don't explain it anywhere, try telnet to LCM1.local (or your LC7001's IP address), port 2112. It will make a connection and then JSON objects will start appearing on your terminal, even if they are just pings. The plugin attempts to decode these objects. Not knowing EVERY possible object that the LC7001 might transmit, I decided to log anything unknown. That way, I could just search the log for these messages and find objects I did not account for. I just searched mine, and this is what I found:
Sep  4 15:58:04 raspberrypi homebridge[250]: [9/4/2019, 3:58:04 PM] [LC7001] Unknown LC7001 initiated service!
Sep  4 15:58:04 raspberrypi homebridge[250]: [9/4/2019, 3:58:04 PM] [LC7001] { ID: 0,Sep  4 15:58:04 raspberrypi homebridge[250]:   Service: 'SystemInfo',
Sep  4 15:58:04 raspberrypi homebridge[250]:   Model: 'LCM1',
Sep  4 15:58:04 raspberrypi homebridge[250]:   FirmwareVersion: '3.1.1-8-g0425',
Sep  4 15:58:04 raspberrypi homebridge[250]:   FirmwareDate: '2018-05-03',
Sep  4 15:58:04 raspberrypi homebridge[250]:   FirmwareBranch: 'LCM1',
Sep  4 15:58:04 raspberrypi homebridge[250]:   MACAddress: 'XX:XX:XX:XX:XX:XX',
Sep  4 15:58:04 raspberrypi homebridge[250]:   HouseID: XXX,
Sep  4 15:58:04 raspberrypi homebridge[250]:   UpdateState:
Sep  4 15:58:04 raspberrypi homebridge[250]:    { Status: 'Downloading',
Sep  4 15:58:04 raspberrypi homebridge[250]:      UpdateVersion: '4:0:1-28',
Sep  4 15:58:04 raspberrypi homebridge[250]:      ReleaseNotes: 'Miscellaneous bug fixes and improvements',
Sep  4 15:58:04 raspberrypi homebridge[250]:      DownloadProgress: 0 },
Sep  4 15:58:04 raspberrypi homebridge[250]:   Status: 'Success' }

Starting with { ID: 0, this is the JSON object. Apparently, I got 101 of these from 0 to 100 for DownloadProgress.... In syslog.1 I found:
Sep  3 15:58:03 raspberrypi homebridge[250]: [9/3/2019, 3:58:03 PM] [LC7001] Unknown LC7001 initiated service!
Sep  3 15:58:03 raspberrypi homebridge[250]: [9/3/2019, 3:58:03 PM] [LC7001] { ID: 0,
Sep  3 15:58:03 raspberrypi homebridge[250]:   Service: 'SystemInfo',
Sep  3 15:58:03 raspberrypi homebridge[250]:   Model: 'LCM1',
Sep  3 15:58:03 raspberrypi homebridge[250]:   FirmwareVersion: '3.1.1-8-g0425',
Sep  3 15:58:03 raspberrypi homebridge[250]:   FirmwareDate: '2018-05-03',
Sep  3 15:58:03 raspberrypi homebridge[250]:   FirmwareBranch: 'LCM1',
Sep  3 15:58:03 raspberrypi homebridge[250]:   MACAddress: 'XX:XX:XX:XX:XX:XX',
Sep  3 15:58:03 raspberrypi homebridge[250]:   HouseID: XXX,
Sep  3 15:58:03 raspberrypi homebridge[250]:   UpdateState:
Sep  3 15:58:03 raspberrypi homebridge[250]:    { Status: 'Downloading',
Sep  3 15:58:03 raspberrypi homebridge[250]:      UpdateVersion: '4:0:0-269',
Sep  3 15:58:03 raspberrypi homebridge[250]:      ReleaseNotes: 'This firmware update is required to migrate your system to Legrand cloud.',
Sep  3 15:58:03 raspberrypi homebridge[250]:      DownloadProgress: 100 },
Sep  3 15:58:03 raspberrypi homebridge[250]:   Status: 'Success' }

So it looks like my LC7001 has been downloading the new firmware, but not updating itself. I can now write a handler for this object and it will stop appearing as an unknown object.
I expect lots of new Unknown objects related to Legrand cloud. Will just have to tackle them as they show up.
Unhandled service: Eliot -- This is more of the same. In a coding decision that might be worth revisiting, I divide incoming JSON objects into two groups. Those that have ID=0, which I called LC7001-initiated, and those where ID != 0, which are responsive objects. If I send an object with ID=X, the LC7001 will respond with an object ID=X. This allows multiple clients to send requests and know which responses are intended for which request. You have received an object with ID!=0 and Service=Eliot. Eliot is probably Legrand cloud (see https://www.legrand.us/eliot.aspx). I would be curious to know what you sent to get an Eliot response, or if LC7001-initiated objects now are not all ID=0, but this just means something is going on that I have no coded for.
If everything appears to be working, then the core JSON interface was not changed and I would expect it to keep working fine. What a relief! These "errors" are in effect informational. They just tell you the LC7001 said something I did not anticipate.
What I did NOT do was throw an error if a known object type has properties I did not anticipate. Consider:
{"ID":0,"Service":"BroadcastDiagnostics","FirmwareVersion":"3.1.1-8g0425","firstSystemTime":1567378655,"MACAddress":"XX:XX:XX","other_send_err:":2,"other_recv_err:":4,"qmotion_try_find:":4623,"task_heartbeat_bitmask:":7,"CurrentTime":1567656180,"uSec":4969,"Seqno":29618,"Status":"Success"}

I handle this by copying the FirmwareVersion, firstSystemTime, MACAddress, and CurrentTime properties into my object. If the firmware were to omit the MACAddress property or change it from a short (3-byte) version to the full 6-byte version, my plugin would do no sanity checking. There would not be any mention that things were amiss in the logs. A project for a different day, perhaps.
Sean
    On Wednesday, September 4, 2019, 7:22:23 PM PDT, Rob Gura <notifications@github.com> wrote:  
 
 
Dang, I wish I would have read this before upgrading my firmware...I guess that makes me the guinea pig.

As of today, I am getting 2 new errors in my homekit console, although it does in face still seem to be actually working. (Odd)

Added the 2 errors as open issues
   
   - Unknown LC7001 initiated object! - As of today (when I upgraded to firmware 4.0.1-28
   - Unhandled service: Eliot Errors - The other day when I upgraded to firmware 4.0.0-269

**Comment by ghost** [comment]
Here is what I see when monitoring via telnet (wow, I have not used that and a loooong time)

pi@raspberrypi:~ $ telnet 192.168.7.67 2112
Trying 192.168.7.67...
Connected to 192.168.7.67.
Escape character is '^]'.
{"ID":0,"Service":"ping","CurrentTime":1567730540,"PingSeq":1,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730545,"PingSeq":2,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730550,"PingSeq":3,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730555,"PingSeq":4,"Status":"Success"}{"Service":"EliotErrors","CurrentTime":1567712558,"EliotConnect":4,"Connect":1,"DNS":3,"UserToken":1}{"ID":0,"Service":"ping","CurrentTime":1567730563,"PingSeq":5,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730568,"PingSeq":6,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730573,"PingSeq":7,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730578,"PingSeq":8,"Status":"Success"}{"ID":0,"Service":"SystemPropertiesChanged","PropertyList":{"CurrentTime":1567730580},"Status":"Success"}{"ID":0,"Service":"BroadcastMemory","FreeMemory:":53756,"FreeMemLowWater:":21884,"Malloc_Count:":615797,"Free_Count:":616728,"JsonConnections:":1,"StaticRamUsage:":40768,"PeakRamUsage:":248659,"CurrentTime":1567730580,"uSec":270,"Seqno":9716,"Status":"Success"}{"ID":0,"Service":"BroadcastDiagnostics","FirmwareVersion":"4.0.1-28-g5ceb","firstSystemTime":1567648641,"MACAddress":"02:D9:CA","CommunicationSequenceNum":5,"other_send_err:":10,"other_recv_err:":17,"qmotion_try_find:":1366,"no_json_msg_slot:":41,"task_heartbeat_bitmask:":3,"CurrentTime":1567730580,"uSec":316,"Seqno":9717,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730585,"PingSeq":9,"Status":"Success"}{"Service":"EliotErrors","CurrentTime":1567712587,"EliotConnect":4,"Connect":1,"DNS":3,"UserToken":1}{"ID":0,"Service":"ping","CurrentTime":1567730592,"PingSeq":10,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730597,"PingSeq":11,"Status":"Success"}^]

telnet> quit
Connection closed.
pi@raspberrypi:~ $

**Comment by sbozarth** [comment]
Of the messages you provided, the BroadcastDiagnostics adds two new properties in your firmware version:
CommunicationSequenceNumno_json_msg_slot
Not sure what these mean; maybe time will tell. I would be interested to know how those numbers change over time.
I published plugin version 0.6.1 that handles the two errors you received. One I already had code for (added 0.4.1), but the code was in the wrong place and thus not catching the object when the Service property was undefined. Moved it outside the conditional that checks for Service being defined. The EliotErrors service is handled by simply printing a stub in the log. Not useful, but it seems like debugging information anyway.
I think I only recently published 0.5.x on NPM which included the auto-reconnect logic. The code was published on github in December, but the NPM package was not updated because I wanted to be the guinea pig for that for a while...then I forgot to ever do it. If your problem with it stopping working is caused by a lost connection, the auto-reconnect should resolve that. 

Sean

    On Thursday, September 5, 2019, 5:47:08 PM PDT, Rob Gura <notifications@github.com> wrote:  
 
 
Here is what I see when monitoring via telnet (wow, I have not used that and a loooong time)

pi@raspberrypi:~ $ telnet 192.168.7.67 2112
Trying 192.168.7.67...
Connected to 192.168.7.67.
Escape character is '^]'.
{"ID":0,"Service":"ping","CurrentTime":1567730540,"PingSeq":1,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730545,"PingSeq":2,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730550,"PingSeq":3,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730555,"PingSeq":4,"Status":"Success"}{"Service":"EliotErrors","CurrentTime":1567712558,"EliotConnect":4,"Connect":1,"DNS":3,"UserToken":1}{"ID":0,"Service":"ping","CurrentTime":1567730563,"PingSeq":5,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730568,"PingSeq":6,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730573,"PingSeq":7,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730578,"PingSeq":8,"Status":"Success"}{"ID":0,"Service":"SystemPropertiesChanged","PropertyList":{"CurrentTime":1567730580},"Status":"Success"}{"ID":0,"Service":"BroadcastMemory","FreeMemory:":53756,"FreeMemLowWater:":21884,"Malloc_Count:":615797,"Free_Count:":616728,"JsonConnections:":1,"StaticRamUsage:":40768,"PeakRamUsage:":248659,"CurrentTime":1567730580,"uSec":270,"Seqno":9716,"Status":"Success"}{"ID":0,"Service":"BroadcastDiagnostics","FirmwareVersion":"4.0.1-28-g5ceb","firstSystemTime":1567648641,"MACAddress":"02:D9:CA","CommunicationSequenceNum":5,"other_send_err:":10,"other_recv_err:":17,"qmotion_try_find:":1366,"no_json_msg_slot:":41,"task_heartbeat_bitmask:":3,"CurrentTime":1567730580,"uSec":316,"Seqno":9717,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730585,"PingSeq":9,"Status":"Success"}{"Service":"EliotErrors","CurrentTime":1567712587,"EliotConnect":4,"Connect":1,"DNS":3,"UserToken":1}{"ID":0,"Service":"ping","CurrentTime":1567730592,"PingSeq":10,"Status":"Success"}{"ID":0,"Service":"ping","CurrentTime":1567730597,"PingSeq":11,"Status":"Success"}^]

telnet> quit
Connection closed.
pi@raspberrypi:~ $

**Comment by ghost** [comment]
> I would be interested to know how those numbers change over time.

Let me know if you need anything specific. For the time being 0.6.1 is installed and seems stable. I am seeing these new console messages:
[9/6/2019, 7:35:38 AM] [LC7001] LC7001 Debug Message: 010301Debug : Connection Log
[9/6/2019, 7:35:38 AM] [LC7001] LC7001 Debug Message: 010302Debug : Connection Log
[9/6/2019, 7:35:38 AM] [LC7001] LC7001 Debug Message: 010303Debug : Connection Log
[9/6/2019, 7:35:38 AM] [LC7001] LC7001 Debug Message: 010304Debug : Connection Log
[9/6/2019, 7:35:38 AM] [LC7001] LC7001 Debug Message: 010305Debug : Connection Log
[9/6/2019, 7:35:39 AM] [LC7001] LC7001 Debug Message: 010306Debug : Connection Log

I am assuming that these are merely informational as everything is working as expected.

**Comment by ghost** [comment]
But wait, there's more. Is the Debug Message number just a sequential number? Any risk of SO once the number exceeds 999999?

[9/6/2019, 7:40:50 AM] [LC7001] LC7001 Debug Message: 010308Debug : Connection Log
[9/6/2019, 7:40:50 AM] [LC7001] LC7001 Debug Message: 010309Debug : Connection Log
[9/6/2019, 7:40:50 AM] [LC7001] LC7001 Debug Message: 010310Debug : Connection Log
[9/6/2019, 7:40:50 AM] [LC7001] LC7001 Debug Message: 010311Debug : Connection Log
[9/6/2019, 7:40:50 AM] [LC7001] LC7001 Debug Message: 010312Debug : Connection Log
[9/6/2019, 7:40:50 AM] [LC7001] LC7001 Debug Message: 010313Debug : Connection Log
[9/6/2019, 7:40:51 AM] [LC7001] LC7001 Debug Message: 010314Debug : Connection Log
[9/6/2019, 7:40:52 AM] [LC7001] LC7001 Debug Message: 010315Debug : Connection Log
[9/6/2019, 7:40:52 AM] [LC7001] LC7001 Debug Message: 010316Debug : Connection Log
[9/6/2019, 7:40:52 AM] [LC7001] LC7001 Debug Message: 010317Debug : Connection Log
[9/6/2019, 7:40:52 AM] [LC7001] LC7001 Debug Message: 010318Debug : Connection Log
[9/6/2019, 7:40:52 AM] [LC7001] LC7001 Debug Message: 010319Debug : Connection Log
[9/6/2019, 7:40:53 AM] [LC7001] LC7001 Debug Message: 010320Debug : Connection Log

**Comment by sbozarth** [comment]
I assume that once they get to 999999 they roll over to 000000, but I don’t know if they have ever gotten that far. If they go to 1000000, the regular expression will not catch them.

Feel free to comment out the line of code that logs debug messages. You are clearly receiving way more than I did when I wrote the handler for it. I just commented out the logging of artik messages (log used to say artik message ignored) in 0.6.1 since my log was now inundated with them and artik is shut down. You probably didn’t have those with new firmware.

Alternative solution is to set a configuration parameter that turns logging of debug messages on or off. Should be easy enough; would just take some time.

Search lc7001.js for “Context” or “Debug”. They should appear on same line that does a regular expression test in an if statement. Comment out the this.log statement immediately following.

Sean

**Comment by sbozarth** [comment]
So, I updated my LC7001 to the new firmware. I did a packet capture before and after, and everything appeared to be the same. Things are stable for you? I'm just getting a lot of EliotErrors messages, so I might comment them out in 0.6.2.
Sean

    On Wednesday, September 4, 2019, 7:22:23 PM PDT, Rob Gura <notifications@github.com> wrote:  
 
 
Dang, I wish I would have read this before upgrading my firmware...I guess that makes me the guinea pig.

As of today, I am getting 2 new errors in my homekit console, although it does in face still seem to be actually working. (Odd)

Added the 2 errors as open issues
   
   - Unknown LC7001 initiated object! - As of today (when I upgraded to firmware 4.0.1-28
   - Unhandled service: Eliot Errors - The other day when I upgraded to firmware 4.0.0-269
